### DISCUSSION TITLE
Debugging pin access

### DISCUSSION THREAD

**Discussion by titan73** [The-OpenROAD-Project/OpenROAD#5464] [Q&A] [answered]
I'm trying to debug pin access put I don't understand well the messages

    [INFO DRT-0119] Marker (23070, 74320) (23260, 74510) on 2:
    Min cut numCuts 1 width 159 cutDistance 710 connection FROMABOVE length -1 distance -1
    [INFO DRT-0119] Marker (23070, 74320) (23260, 74510) on 4:
    Min cut numCuts 1 width 159 cutDistance 710 connection FROMBELOW length -1 distance -1

What is "2:" and "4:"?

I don't undestand the line:

    Min cut numCuts 1 width 159 cutDistance 710 connection FROMABOVE length -1 distance -1

<img width="294" alt="via" src="https://github.com/user-attachments/assets/cf7e3c06-9f4d-4a8c-a9dc-f4ef58c595a9">

**Comment by maliberty** [comment, marked as answer]
2 & 4 are the layer numbers.  The line is the rule that is violated in this attempt at pa.

**Comment by titan73** [comment]
Ah ok. I found them with the color but yeah got it 1 => contact, 2 => m1, 3 => v1 and 4 => m2. Thanks!
That said, shouldn't this be more straightforward to use the layer name instead?

**Comment by titan73** [comment]
I didn't understand the values in:

    Min cut numCuts 1 width 159 cutDistance 710 connection FROMABOVE length -1 distance -1

An particularly with the -1 but looking at the techno lef I have:

	MINIMUMCUT 1 WIDTH 0.159 WITHIN 0.71  FROMABOVE  ;
	MINIMUMCUT 2 WIDTH 1.4 WITHIN 0.71  FROMABOVE  ;

LENGTH length WITHIN distance are not specified thus the -1.

The metal1 vertical pin the router is trying to check accessibility is 304 x 619. The router seems to check pin access from metal2.

The shapes dimensions we can see are:
    - metal1 => 290 x 210
    - metal2 => 200 x 290

I don't understand what's wrong and what the rule applies to i.e the pin or the temporary shapes set by the router debug.

Any idea?

**Comment by titan73** [comment]
Ah the shapes are from the via rules. It's not easy to figure out which via rule it is exactly as it is not mentioned in the debug message (that would be nice to have it) but I retrieved it. So the via is 190 x 190.

So basically the width of the shape in metal1 is 210 and in metal2 it's 200. It's > 159 so there should be 1 cut via if I understand well. Still not seeing what's wrong though.

**Comment by maliberty** [comment]
Its pretty hard to offer help without being able to see the data.  Is it possible to make a one gate test case with some sanitized values you can share?

**Reply by titan73** [reply]
I'll try to make one.

**Comment by titan73** [comment]
Here is a test case.

[pin_access.zip](https://github.com/user-attachments/files/16416815/pin_access.zip)

**Comment by maliberty** [comment]
What is the purpose of
```
	MINIMUMCUT 1 WIDTH 0.285 WITHIN 1.270  FROMBELOW  ;
```

If I remove it then PA passes (though a later issue shows up).  You can't have less than one cut.

**Reply by titan73** [reply]
Regarding WIDTH, I'd say to have 1 cut via, a minimum width has to be present but for WITHIN I don't know.
I'll ask the team responsible for this techno and let you know.
Note that all values are changed but the aspect ratio is kept.

**Reply by maliberty** [reply]
At any width you still need one cut, no?

**Reply by titan73** [reply]
Yes. You're right. A colleague from CAD told me this line is indeed useless.
I asked people from the techno but the answer will take time as they are in vacations.
For the next issue:
    [INFO DRT-0181] Start track assignment.
    Error: test.tcl, 26 vector::_M_range_check: __n (which is 4) >= this->size() (which is 0)

Should I enter a bug?

**Reply by maliberty** [reply]
Yes

**Reply by titan73** [reply]
Ok done. Thanks!
